traffic_ops/traffic_ops_golang/deliveryservice/deliveryservice_server_capabilities.go

```diff
@@ -0,0 +1,196 @@
+func (sc *ServerCapability) ParamColumns() map[string]dbhelpers.WhereColumnInfo {
+	return map[string]dbhelpers.WhereColumnInfo{
+		deliveryServiceQueryParam: dbhelpers.WhereColumnInfo{
+			Column:  "sc.deliveryservice_id",
+			Checker: api.IsInt,
+		},
+		xmlIDQueryParam: dbhelpers.WhereColumnInfo{
+			Column:  "ds.xml_id",
+			Checker: nil,
+		},
+		serverCapabilityQueryParam: dbhelpers.WhereColumnInfo{
+			Column:  "sc.server_capability",
+			Checker: nil,
+		},
+	}
+}
+
+// DeleteQuery implements the api.GenericDeleter interface
+func (sc *ServerCapability) DeleteQuery() string {
+	return `DELETE FROM deliveryservice_server_capability
+	WHERE deliveryservice_id = :deliveryservice_id AND server_capability = :server_capability`
+}
+
+// GetKeyFieldsInfo implements the api.Identifier interface
+func (sc ServerCapability) GetKeyFieldsInfo() []api.KeyFieldInfo {
+	return []api.KeyFieldInfo{
+		{
+			Field: deliveryServiceQueryParam,
+			Func:  api.GetIntKey,
+		},
+		{
+			Field: xmlIDQueryParam,
+			Func:  api.GetStringKey,
+		},
+		{
+			Field: serverCapabilityQueryParam,
+			Func:  api.GetStringKey,
+		},
+	}
+}
+
+// GetKeys implements the api.Identifier interface and is not needed
+// because Update is not available
+func (sc ServerCapability) GetKeys() (map[string]interface{}, bool) {
+	return nil, false
+}
+
+// SetKeys implements the api.Identifier interface and allows the
+// create handler to assign deliveryServiceID and serverCapability.
+func (sc *ServerCapability) SetKeys(keys map[string]interface{}) {
+	// this utilizes the non panicking type assertion, if the thrown
+	// away ok variable is false it will be the zero of the type
+	id, _ := keys[deliveryServiceQueryParam].(int)
+	sc.DeliveryServiceID = &id
+
+	capability, _ := keys[serverCapabilityQueryParam].(string)
+	sc.ServerCapability = &capability
+}
+
+// GetAuditName implements the api.Identifier interface and
+// returns the name of the object.
+func (sc *ServerCapability) GetAuditName() string {
+	if sc.ServerCapability != nil {
+		return *sc.ServerCapability
+	}
+	return "unknown"
+}
+
+// GetType implements the api.Identifier interface and
+// returns the name of the struct
+func (sc *ServerCapability) GetType() string {
+	return "deliveryservice.ServerCapability"
+}
+
+// Validate implements the api.Validator interface
+func (sc ServerCapability) Validate() error {
+	errs := validation.Errors{
+		deliveryServiceQueryParam:  validation.Validate(sc.DeliveryServiceID, validation.Required),
+		serverCapabilityQueryParam: validation.Validate(sc.ServerCapability, validation.Required),
+	}
+
+	return util.JoinErrs(tovalidate.ToErrors(errs))
+}
+
+// Update implements the api.CRUDer interface
+func (sc *ServerCapability) Update() (error, error, int) {
+	return nil, nil, http.StatusNotImplemented
+}
+
+// Read implements the api.CRUDer interface
+func (sc *ServerCapability) Read() ([]interface{}, error, error, int) {
+	return api.GenericRead(sc)
```

Comment:
This isn't checking Tenancy on the "objects"' Delivery Service(s). Is this API not meant to respect Tenancy? Because it ought to - at the very least if an endpoint gives away that a Delivery Service exists it's circumventing the protection provided by Tenancy checks on the `/deliveryservices` endpoint.

Reply:
@ocket8888 I will look into this, this is my first PR against the go-api. I havent heard of this yet.

Reply:
@ocket8888 thanks for the comment, I wasnt familiar with this. I will update the code to check tenancy.